drivers/spi/opentitan/spi.c

```diff
@@ -0,0 +1,727 @@
+        regs->TXDATA = words[i];
+    }
+
+    // Transmit trailing bytes
+    uint8_t *trailing_bytes = buffer + num_leading_bytes + num_words * sizeof(uint32_t);
+    uint16_t num_trailing_bytes = (len - num_leading_bytes) % sizeof(uint32_t);
+
+    for (uint16_t i = 0; i < num_trailing_bytes; i++) {
+        *TX_BYTE = trailing_bytes[i];
+    }
+}
+
+int recieve_data(void *buffer, uint16_t len) {
+    // Recieve leading bytes to word align buffer
+    uint8_t *leading_bytes = buffer;
+    const uint16_t bytes_until_aligned = ALIGN(buffer, sizeof(uint32_t)) - buffer;
+    const uint16_t num_leading_bytes = MIN(len, bytes_until_aligned); 
+
+    // Treat rx_data as queue
+    uint64_t rx_data = 0;
+    uint8_t *rx_byte = (uint8_t *) &rx_data;
+
+    if (num_leading_bytes > 0) {
+        if (poll_rx()) {
+            return -1;
+        }
+        rx_data = regs->RXDATA;
+
+        for (uint16_t i = 0; i < num_leading_bytes; i++) {
+            leading_bytes[i] = rx_byte[i];
+        } 
+        rx_data >>= num_leading_bytes * 8;
+
+        bool done = num_leading_bytes != bytes_until_aligned;
+        if (done) {
+            return 0;
+        }
+    }
+
+    // Transmit as many words as possible
+    uint32_t *words = buffer + num_leading_bytes;
+    const uint16_t num_words = (len - num_leading_bytes) / sizeof(uint32_t);
+
+    for (uint16_t i = 0; i < num_words; i++) {
+        if (poll_rx()) {
+            return -1;
+        }
+
+        rx_data |= regs->RXDATA << (sizeof(uint32_t) - num_leading_bytes) * 8; 
+        words[i] = rx_data;
+        rx_data >>= sizeof(uint32_t) * 8;
+    }
+
+    // Transmit trailing bytes
+    uint8_t *trailing_bytes = buffer + num_leading_bytes + num_words * sizeof(uint32_t);
+    const uint16_t num_trailing_bytes = (len - num_leading_bytes) % sizeof(uint32_t);
+
+    if (num_trailing_bytes == 0) {
+        return 0;
+    }
+    
+    if (poll_rx()) {
+        return -1;
+    }
+    rx_data |= regs->RXDATA << (sizeof(uint32_t) - num_leading_bytes) * 8;
+    for (uint16_t i = 0; i < num_trailing_bytes; i++) {
+        trailing_bytes[i] = rx_byte[i];
+    }
+    
+    return 0;
+}
+
+void state_idle(void) {
+    spi_reset_state(&driver_data);
+
+    if (!spi_queue_empty(queue_handle.request->ctrl)) {
+        fsm_state.nxt_state = REQ;
+    }
+    else { 
+        fsm_state.yield = true;
+    }
+
+    return;
+}
+
+void state_req(void) {
+    int err = spi_dequeue_request(
+        queue_handle,
+        &driver_data.cs_line,
+        (uintptr_t *) &driver_data.cmd,
+        (uintptr_t *) &driver_data.slice_base, 
+        &driver_data.num_cmds
+    );
+
+    if (err) {
+        LOG_DRIVER_ERR("fatal: failed to dequeue request\n"); 
+        // Do not respond, as a failed dequeue implies there is no request 
+        // to respond to
+        fsm_state.nxt_state = IDLE;
+        return;
+    }
+
+    // Check if the amount of commands make sense
+    if (driver_data.num_cmds > SPI_CMD_CAPACITY) {
```

Comment:
Why is num_cmds a uint16_t when it can fit into a uint8_t? As mentioned in the comment on the header file, better to resize this to match the other 8-bit sizes that are used similarly